src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -214,7 +216,12 @@ fn compile<'a, 'cfg>(targets: &[(&'a Target, &'a Profile)],
         drop(profiling_marker);
 
         // Be sure to compile all dependencies of this target as well.
-        for &(pkg, target, p) in cx.dep_targets(pkg, target, profile).iter() {
+        let all_targets: HashSet<(&Package, &Target, &Profile)> =
+            kinds.iter().flat_map(|&kind| {
+                cx.dep_targets(pkg, target, kind, profile).into_iter()
+            }).collect();
```

Comment:
Hm, these extra loops and sets make me think that there may be a better route here. It may be the case that this function should take `&[(Target, Profile, Kind)]` (e.g. add a `Kind`) and then it won't have to dispatch based on `get_requirement` in the `rustc` case.

Overall I'm starting to think that `get_requirement` may want to be phased out entirely as it seems like a somewhat broken notion now, but I'm also not 100% sure of what the impact of such a change would be.